File: python/VCFOneParentPhased.py
from __future__ import annotations

# coding: utf-8
# VCFOneParentPhased.py
# 片方の親だけがphasingされた家系

from collections import defaultdict, Counter
from typing import List, Tuple, Optional, IO, Dict, Iterator

from VCFFamily import *
from VCFImputable import *
from Map import *
import Imputer
from option import *


#################### VCFHOneParentPhased ####################

class VCFOneParentPhased(VCFBase, VCFFamilyBase, VCFImputable):
	def __init__(self, header: list[list[str]],
						records: list[VCFFamilyRecord],
						mat_p: bool, map_: Map, ref_vcf: VCFSmall):
		VCFBase.__init__(self, header)
		VCFFamilyBase.__init__(self)
		VCFImputable.__init__(self, map_)
		self.records: list[VCFFamilyRecord] = records
		self.is_mat_phased: bool = mat_p
		self.ref_vcf: VCFSmall = ref_vcf
	
	def get_header(self) -> list[list[str]]:
		return self.header
	
	def get_samples(self) -> list[str]:
		return self.samples
	
	def __len__(self) -> int:
		return len(self.records)
	
	def get_record(self, i: int) -> VCFRecord:
		return self.records[i]
	
	def get_family_record(self, i: int) -> VCFFamilyRecord:
		return self.records[i]
	
	def clip_ref_haplotype(self, sample_id: int, side: int) -> Haplotype:
		hap = self.ref_vcf.clip_raw_haplotype(sample_id, side)
		return Haplotype(hap, sample_id, side)
	
	# divide VCF by 1 cM
	def divide_by_cM(self) -> Iterator[VCFOneParentPhased]:
		# 1cMを超えても10個以内で10cM以内なら塊とみなす
		rs = self.records[:1]
		ref_rs = self.ref_vcf.records[:1]
		for i in range(1, len(self)):
			record = self.records[i]
			ref_record = self.ref_vcf.records[i]
			if self.is_block(record, rs):
				rs.append(record)
				ref_rs.append(ref_record)
			else:
				ref_vcf_cM = VCFSmall(self.ref_vcf.header, ref_rs)
				yield VCFOneParentPhased(self.header, rs, self.is_mat_phased,
														self.map, ref_vcf_cM)
				rs = [record]
				ref_rs = [ref_record]
		
		ref_vcf_cM = VCFSmall(self.ref_vcf.header, ref_rs)
		yield VCFOneParentPhased(self.header, rs, self.is_mat_phased,
														self.map, ref_vcf_cM)
	
	def collect_haplotypes_from_parents(self) -> list[Haplotype]:
		i = 0 if self.is_mat_phased else 1
		return [ self.clip_haplotype(i, j) for j in (0, 1) ]
	
	def collect_haplotype_from_refs(self) -> list[Haplotype]:
		return [ self.clip_ref_haplotype(i, j)
					for i in range(self.ref_vcf.num_samples())
					for j in (0, 1) ]
	
	def collect_haplotypes_mat(self, sample_index: int) -> list[Haplotype]:
		if self.is_mat_phased:
			return self.collect_haplotypes_from_parents()
		else:
			return self.collect_haplotype_from_refs()
	
	def collect_haplotypes_pat(self, sample_index: int) -> list[Haplotype]:
		if self.is_mat_phased:
			return self.collect_haplotype_from_refs()
		else:
			return self.collect_haplotypes_from_parents()
	
	def impute_cM(self, prev_haps: list[HaplotypePair]) -> list[HaplotypePair]:
		haps: list[HaplotypePair] = []
		for sample_index, prev_hap in enumerate(prev_haps, 2):
			hap = self.impute_cM_each_sample(prev_hap, sample_index, True)
			haps.append(hap)
		return haps
	
	def impute(self):
		if not self.records:
			return
		
		h = Haplotype.default_value()
		haps = [(h, h)] * (self.num_progenies())
		for vcf_cM in self.divide_by_cM():
			haps = vcf_cM.impute_cM(haps)
	
	@staticmethod
	def create(samples: list[str], is_mat_phased: bool,
				imputed_vcf: VCFSmallBase, orig_vcf: VCFSmallBase,
				gmap: Map, ref_vcf: VCFSmall) -> VCFOneParentPhased:
		vcf = VCFFamily.create_by_two_vcfs(imputed_vcf, orig_vcf, samples)
		new_vcf = VCFOneParentPhased(vcf.header, vcf.records,
											is_mat_phased, gmap, ref_vcf)
		return new_vcf
	
	# samplesに対応するRecordを作る
	@staticmethod
	def merge_records(vcfs: list[VCFFamilyBase], i: int, samples: list[str],
					dic_pos: Dict[str, tuple[VCFFamilyBase, int]]) -> VCFRecord:
		v = vcfs[0].get_family_record(i).v[:9]
		for sample in samples:
			vcf, c = dic_pos[sample]
			v.append(vcf.get_family_record(i).v[c])
		return VCFRecord(v, samples)

__all__ = ['VCFOneParentPhased']

File: src/VCFOneParentPhased.cpp
#include <stack>
#include <sstream>
#include <algorithm>
#include <cmath>
#include <random>
#include <cassert>
#include "../include/common.h"
#include "../include/VCFOneParentPhased.h"
#include "../include/VCFFillable.h"
#include "../include/Map.h"
#include "../include/Pedigree.h"
#include "../include/Imputer.h"

using namespace std;


//////////////////// VCFOneParentPhased ////////////////////

VCFOneParentPhased::VCFOneParentPhased(const vector<STRVEC>& h, const STRVEC& s,
									   vector<VCFFamilyRecord *> rs, bool mat_p,
									   const Map& m, const VCFSmall *ref) :
						VCFBase(h, s), VCFFamilyBase(), VCFImputable(m),
						records(rs), is_mat_phased(mat_p), ref_vcf(ref) { }

VCFOneParentPhased::~VCFOneParentPhased() {
	for(auto p = records.begin(); p != records.end(); ++p)
		delete *p;
}

bool VCFOneParentPhased::is_mat_hetero() const {
	return records.front()->is_hetero(0);
}

Haplotype VCFOneParentPhased::clip_haplotype(
									size_t sample_id, int i) const {
	const auto	hap = clip_raw_haplotype(sample_id, i);
	return Haplotype(hap, sample_id, i);
}

Haplotype VCFOneParentPhased::clip_ref_haplotype(
									size_t sample_id, int i) const {
	const auto	hap = ref_vcf->clip_raw_haplotype(sample_id, i);
	return Haplotype(hap, sample_id, i);
}

VCFOneParentPhased *VCFOneParentPhased::divide_by_positions(
											size_t first, size_t last) const {
	vector<VCFFamilyRecord *>	rs(records.begin() + first,
								   records.begin() + last);
	vector<VCFRecord *>	ref_rs(ref_vcf->get_records().begin() + first,
							   ref_vcf->get_records().begin() + last);
	const VCFSmall	*ref_vcf_divided = new VCFSmall(ref_vcf->get_header(),
													ref_vcf->get_samples(),
													ref_rs, true);
	return new VCFOneParentPhased(header, samples, rs, is_mat_phased, get_map(),
															ref_vcf_divided);
}

vector<Haplotype> VCFOneParentPhased::collect_haplotypes_from_parents() const {
	const size_t	i = this->is_mat_phased ? 0 : 1;
	const vector<Haplotype>	haps { clip_haplotype(i, 0), clip_haplotype(i, 1) };
	return haps;
}

vector<Haplotype> VCFOneParentPhased::collect_haplotype_from_refs() const {
	vector<Haplotype>	haps;
	for(size_t i = 0; i < ref_vcf->num_samples(); ++i) {
		haps.push_back(clip_ref_haplotype(i, 0));
		haps.push_back(clip_ref_haplotype(i, 1));
	}
	return haps;
}

vector<Haplotype> VCFOneParentPhased::collect_haplotypes_mat(
												size_t sample_index) const {
	if(this->is_mat_phased)
		return collect_haplotypes_from_parents();
	else
		return collect_haplotype_from_refs();
}

vector<Haplotype> VCFOneParentPhased::collect_haplotypes_pat(
												size_t sample_index) const {
	if(this->is_mat_phased)
		return collect_haplotype_from_refs();
	else
		return collect_haplotypes_from_parents();
}

vector<HaplotypePair> VCFOneParentPhased::impute_cM(
									const vector<HaplotypePair>& prev_haps) {
	vector<HaplotypePair>	haps;
	for(size_t i = 0; i < prev_haps.size(); ++i) {
		const auto	prev_hap = prev_haps[i];
		const auto	hap = this->impute_cM_each_sample(prev_hap, i+2,
															true, true);
		haps.push_back(hap);
	}
	return haps;
}

void VCFOneParentPhased::impute() {
	const Haplotype	h = Haplotype::default_value();
	vector<HaplotypePair>	haps(num_progenies(), make_pair(h, h));
	const auto	vcf_cMs = VCFImputable::divide_by_cM(this);
	for(auto p = vcf_cMs.begin(); p != vcf_cMs.end(); ++p) {
		VCFOneParentPhased	*vcf_cM = *p;
		haps = vcf_cM->impute_cM(haps);
		vcf_cM->records.clear();
		delete vcf_cM->ref_vcf;
		delete vcf_cM;
	}
}

VCFOneParentPhased *VCFOneParentPhased::create(
						const STRVEC& samples, bool is_mat_phased,
						const VCFSmall *merged_vcf, const VCFSmall *orig_vcf,
						const Map& gmap, const VCFSmall *ref_vcf) {
	VCFFamily	*vcf = VCFFamily::create_by_two_vcfs(merged_vcf,
														orig_vcf, samples);
	vector<VCFFamilyRecord *>	records = vcf->get_family_records();
	auto	new_vcf = new VCFOneParentPhased(vcf->get_header(),
												vcf->get_samples(), records,
												is_mat_phased, gmap, ref_vcf);
	vcf->clear_records();
	delete vcf;
	return new_vcf;
}

void VCFOneParentPhased::impute_in_thread(void *config) {
	const auto	*c = (VCFOneParentPhased::ConfigThread *)config;
	for(size_t i = c->first; i < c->size(); i += c->num_threads) {
		VCFOneParentPhased	*vcf = c->vcfs[i];
		vcf->impute();
	}
}

void VCFOneParentPhased::impute_in_parallel(
					const vector<VCFOneParentPhased *>& vcfs, int num_threads) {
	const int	T = min((int)vcfs.size(), num_threads);
	vector<ConfigThread *>	configs(T);
	for(int i = 0; i < T; ++i)
		configs[i] = new ConfigThread(vcfs, (size_t)i, T);
	
#ifndef DEBUG
	vector<pthread_t>	threads_t(T);
	for(int i = 0; i < T; ++i)
		pthread_create(&threads_t[i], NULL,
						(void *(*)(void *))&impute_in_thread,
						(void *)configs[i]);
	
	for(int i = 0; i < T; ++i)
		pthread_join(threads_t[i], NULL);
#else
	for(int i = 0; i < T; ++i)
		impute_in_thread(configs[i]);
#endif
	
	Common::delete_all(configs);
}

STRVEC VCFOneParentPhased::collect_samples(
								const vector<VCFOneParentPhased *>& vcfs) {
	set<string>	set_samples;
	for(auto p = vcfs.begin(); p != vcfs.end(); ++p) {
		const auto	*vcf = *p;
		if(vcf->is_mat_phased)
			set_samples.insert(vcf->pat());
		else
			set_samples.insert(vcf->mat());
		for(size_t i = 2; i < vcf->num_samples(); ++i)
			set_samples.insert(vcf->samples[i]);
	}
	return STRVEC(set_samples.begin(), set_samples.end());
}

// make a record corresponding to samples
VCFSmall *VCFOneParentPhased::merge(const vector<VCFOneParentPhased *>& vcfs) {
	const STRVEC	samples = collect_samples(vcfs);
	
	// Where sample is located in which VCF
	map<string, pair<const VCFOneParentPhased *, size_t>>	dic_pos;
	for(auto p = vcfs.begin(); p != vcfs.end(); ++p) {
		const VCFOneParentPhased	*vcf = *p;
		for(size_t k = 0; k < vcf->num_samples(); ++k)
			dic_pos[vcf->get_samples()[k]] = make_pair(vcf, k + 9);
	}
	
	// create records from the above dictionary
	vector<VCFRecord *>	records;
	for(size_t i = 0; i < vcfs.front()->size(); ++i) {
		const STRVEC&	v1 = vcfs.front()->get_record(i)->get_v();
		STRVEC	v(v1.begin(), v1.begin() + 9);
		for(auto p = samples.begin(); p != samples.end(); ++p) {
			auto	q = dic_pos[*p];
			v.push_back(q.first->get_record(i)->get_v()[q.second]);
		}
		records.push_back(new VCFRecord(v, samples));
	}
	vector<STRVEC>	header = vcfs.front()->trim_header(samples);
	return new VCFSmall(header, samples, records);
}

VCFSmall *VCFOneParentPhased::impute_all(
					const vector<VCFOneParentPhased *>& vcfs, int num_threads) {
	impute_in_parallel(vcfs, num_threads);
	return merge(vcfs);
}
